Python/Product/Debugger/DebugAdapterRemoteProcess.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Debugger.DebugAdapterHost.Interfaces;
+using Microsoft.PythonTools.Infrastructure;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.PythonTools.Debugger {
+    sealed class DebugAdapterRemoteProcess : ITargetHostProcess, IDisposable {
+        private const int _debuggerConnectionTimeout = 5000; // 5000 ms
+        private Stream _stream;
+        private bool _debuggerConnected = false;
+
+        private DebugAdapterRemoteProcess() {}
+
+        public static ITargetHostProcess Attach(string attachJson) {
+            var process = new DebugAdapterRemoteProcess();
+            process.AttachProcess(attachJson);
+            return process;
+        }
+
+        private void AttachProcess(string attachJson) {
+            var json = JObject.Parse(attachJson);
+            var uri = new Uri(json["remote"].Value<string>());
+            ConnectSocket(uri);
+        }
+
+        private void ConnectSocket(Uri uri) {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.IP);
+            EndPoint endpoint;
+            if (uri.IsLoopback) {
+                endpoint = new IPEndPoint(IPAddress.Loopback, uri.Port);
+            } else {
+                endpoint = new DnsEndPoint(uri.Host, uri.Port);
+            }
+            Debug.WriteLine("Connecting to remote debugger at {0}", uri.ToString());
+            var connection = Task.Factory.FromAsync(
+                (AsyncCallback callback, object state) => socket.BeginConnect(endpoint, callback, state),
+                socket.EndConnect,
+                null);
+
+            try {
+                if (connection.Wait(_debuggerConnectionTimeout)) {
```

Comment:
This should probably use whatever machinery PTVS has to do blocking waits on async tasks in a way that prevents deadlocks and similar, since we don't know which thread this is going to be called on.

Reply:
I will try with `TaskHelpers.RunSynchronouslyOnUIThread`. Let me know if this is not the right thing to use here.
```
TaskHelpers.RunSynchronouslyOnUIThread(ct => Task.WhenAny(
                    Task.Factory.FromAsync(socket.BeginConnect, socket.EndConnect, endpoint, null), 
                    Task.Delay(_debuggerConnectionTimeout), 
                    Task.FromCanceled(ct)));
```

Reply:
That's overkill - we don't need the UI thread here really, we just need to wait safely. @AlexanderSher, what's the safe `Wait` equivalent that goes through VS joinable task factory machinery?

Reply:
`JoinableTaskFactory.Run`. Why can't you make `DebugAdapterRemoteProcess.Attach` method async?

Reply:
If you have a service provider, `_site.GetUIThread().InvokeTaskSync(...)`. If not, use `ThreadHelper.JoinableTaskFactory.Run()`. (Don't use the global service provider without also having a fallback to `ThreadHelper`, and in general we strongly avoid using the global service provider.)

Reply:
`_site.GetUIThread().InvokeTaskSync` will also use UI thread, which isn't required here.